byron/ledger/executable-spec/src/Ledger/Delegation.hs

```diff
@@ -478,37 +519,41 @@ instance Embed ADELEGS DELEG where
 -- Generators
 --------------------------------------------------------------------------------
 
--- | Generate delegation certificates, using the allowed delegators in the
--- environment passed as parameter.
-dcertGen :: DSEnv -> Gen DCert
-dcertGen env = do
-  -- The generated delegator must be one of the genesis keys in the
-  -- environment.
-  vkS <- Gen.element $ Set.toList (env ^. allowedDelegators)
-  vkD <- vkGen
-  let Epoch n = env ^. epoch
-  m   <- Gen.integral (linear 0 100)
-  let epo = Epoch (n + m)
-  return DCert { _dbody = (vkD, epo)
-               , _dwit = Sig vkS (owner vkS)
-               , _dwho = (vkS, vkD)
-               , _depoch = epo
-               }
-
--- | Generate a list of delegation certificates.
---
--- At the moment the size of the generated list is severely constrained since
--- the longer the list the higher the probability that it will contain
--- conflicting delegation certificates (that will be rejected by the
--- transition-system-rules).
-dcertsGen :: DSEnv -> Gen [DCert]
--- NOTE: at the moment we cannot use a linear range that depends on the
--- generator size: the problem is that the more delegation certificates in the
--- resulting list, the higher the probability that this list will be rejected
--- and the generator will have to retry.
---
-dcertsGen env = Gen.frequency
-  [(95, pure []), (10, Gen.list (constant 1 n) (dcertGen env))]
+dcertGen :: DSEnv -> DIState -> Gen (Maybe DCert)
+dcertGen env st =
+  let
+    allowed :: [VKeyGenesis]
+    allowed = Set.toList (_dSEnvAllowedDelegators env)
+    -- We can generate delegation certificates using the allowed delegators,
+    -- and we can delegate for the current or next epoch only.
+    preCandidates :: Set (Epoch, VKeyGenesis)
+    preCandidates = Set.fromList
+                  $  zip (repeat $ _dSEnvEpoch env) allowed
+                  ++ zip (repeat $ _dSEnvEpoch env + 1) allowed
+    -- We obtain the candidates by removing the ones that already delegated in
+    -- this or next epoch.
+    candidates :: [(Epoch, VKeyGenesis)]
+    candidates = Set.toList $ preCandidates \\ eks st
+    -- Next, we choose to whom these keys delegate.
+    target :: [VKey]
+    -- NOTE: we might want to make this configurable for now we chose an upper
+    -- bound equal to two times the number of genesis keys to increase the
+    -- chance of having two genesis keys delegating to the same key.
+    target = VKey . Owner <$> [0 .. (2 * fromIntegral (length allowed))]
+
+    mkDCert' ((e, vkg), vk) = DCert (vk, e) (Sig vkg (owner vkg)) (vkg, vk) e
+  in
+
+  if null candidates
+    then return Nothing
+    else Just <$> Gen.element (mkDCert' <$> zip candidates target)
+
+
+dcertsGen :: DSEnv -> DIState -> Gen [DCert]
+dcertsGen env st =
+  -- NOTE: alternatively we could define a HasTrace instance for SDELEG and use
+  -- trace here.
+  catMaybes <$> Gen.list (constant 1 n) (dcertGen env st)
```

Comment:
This can result in an empty list, can't it? Is that fine for this generator? The part that says `constant 1 n` might allude the reader to think it is a non-empty list.

Reply:
Good point. I added the following comment:

```haskell
  -- This generator can result in an empty list of delegation certificates if
  -- no delegation certificates can be produced, according to the delegation
  -- rules, given the initial state and environment.
```